test/StreamJsonRpc.Tests/MarshalableProxyTests.cs

```diff
@@ -499,6 +585,154 @@ public async Task DisposeOnDisconnect()
+    {
+        this.server.ReturnedMarshalableWithSubTypes = new MarshalableSubType2();
+        IMarshalableSubType2? proxy = (IMarshalableSubType2?)await this.client.GetMarshalableWithSubTypesAsync();
+        Assert.Equal(1, await proxy!.GetAsync(1));
+        Assert.Equal(3, await proxy.GetPlusTwoAsync(1));
+        Assert.False(proxy is IMarshalableSubType1);
+        Assert.False(proxy is IMarshalableSubType2Extended);
+    }
+
+    [Fact]
+    public async Task RpcMarshalableKnownSubType_UnknownSubType()
+    {
+        this.server.ReturnedMarshalableWithSubTypes = new MarshalableUnknownSubType();
+        IMarshalableWithSubTypes? proxy = await this.client.GetMarshalableWithSubTypesAsync();
+        Assert.Equal(1, await proxy!.GetAsync(1));
+        Assert.False(proxy is IMarshalableSubType1);
+        Assert.False(proxy is IMarshalableSubType2);
+        Assert.False(proxy is IMarshalableSubType2Extended);
+    }
+
+    [Fact]
+    public async Task RpcMarshalableKnownSubType_OnlyAttibutesOnDeclaredTypeAreHonored()
+    {
+        this.server.ReturnedMarshalableWithSubTypes = new MarshalableSubType2Extended();
+        IMarshalableSubType2? proxy = (IMarshalableSubType2?)await this.client.GetMarshalableWithSubTypesAsync();
+        Assert.Equal(1, await proxy!.GetAsync(1));
+        Assert.Equal(3, await proxy.GetPlusTwoAsync(1));
+        Assert.False(proxy is IMarshalableSubType2Extended);
+    }
+
+    [Fact]
+    public async Task RpcMarshalableKnownSubType_MarshalableNonExtendingBase()
+    {
+        this.server.ReturnedMarshalableWithSubTypes = new MarshalableNonExtendingBase();
+        IMarshalableWithSubTypes? proxy = await this.client.GetMarshalableWithSubTypesAsync();
+        Assert.Equal(1, await proxy!.GetAsync(1));
+
+        Assert.Equal(5, await ((IMarshalableNonExtendingBase)proxy).GetPlusFourAsync(1));
+    }
+
+    [Fact]
+    public async Task RpcMarshalableKnownSubType_IntermediateNonMarshalableInterface()
+    {
+        this.server.ReturnedMarshalableWithSubTypes = new MarshalableSubTypeWithIntermediateInterface();
+        IMarshalableWithSubTypes? proxy = await this.client.GetMarshalableWithSubTypesAsync();
+        Assert.Equal(1, await proxy!.GetAsync(1));
+
+        Assert.Equal(1, await ((IMarshalableSubTypeIntermediateInterface)proxy).GetAsync(1));
+        Assert.Equal(2, await ((IMarshalableSubTypeIntermediateInterface)proxy).GetPlusOneAsync(1));
+
+        // This should return 3, since MarshalableSubTypeWithIntermediateInterface implements this method explicitly
+        // but StreamJsonRpc doesn't know about the IMarshalableSubTypeIntermediateInterface because it is not part
+        // of the RPC contract, so IMarshalableSubTypeWithIntermediateInterface.GetPlusTwoAsync is invoked instead.
+        Assert.Equal(-3, await ((IMarshalableSubTypeIntermediateInterface)proxy).GetPlusTwoAsync(1));
```

Comment:
@AArnott, I want to bring your attention to this.
I think I covered, in the next two unit tests, all other combinations of objects implementing multiple interfaces with implicit and explicit implementations and the use of the `new` keyword in interfaces.

This is the only case I could find that doesn't behave well. But the problem is already present in the current implementation (even the main service proxy is affected): if you implicitly implement a hidden method from a base interface that is not part of the Rpc contract (is not one of the known sub-types), we cannot dispatch the call to that method because it doesn't have a name. The call is dispatched to the wrong method instead.

I will add a detailed explanation of this issue in the docs and provide a general guidance to users to avoid, when possible, using the `new` keyword and method name conflicts when building Rpc interfaces.

I think it is good to have this weird behavior tested here, even if it is less than ideal, so that we at least keep this behavior constant.

Reply:
> in the next two unit tests, all other combinations of objects implementing multiple interfaces with implicit and explicit implementations and the use of the `new` keyword in interfaces.

Thank you! Exhaustive testing of this complex area is super important both for correctness and to maintain such subtle behaviors over time.

> I think it is good to have this weird behavior tested here, even if it is less than ideal, so that we at least keep this behavior constant.

💯 I absolutely agree with your reasoning on that point.

> if you implicitly implement a hidden method from a base interface that is not part of the Rpc contract (is not one of the known sub-types), we cannot dispatch the call to that method because it doesn't have a name.

This part I don't understand. How can you *implicitly* implement an interface method without your method having a name? I understand that `public void Hello() { }` may implicitly implement an interface method, but that method has a name, and it _should_ be available to dispatch RPC requests to.
An *explicit* interface implementation on the other hand (e.g. `void Iface.Hello()`) may be a different matter. I don't remember whether we add these as target methods without the caller specifically asking for that interface to be added. Is that what you're talking about?

Reply:
When I wrote "it doesn't have a name", I explained myself really badly.
I should have written

> we cannot dispatch the call to that method because it is neither declared nor inherited by one of the Rpc interfaces: methods from base interfaces are usually inherited by the Rpc interface but, in this case, the method is hidden.

I think I have clearly understandable examples for this issue that I added to the docs (it's much easier to reason on this with an example). We can have a call and go through this if you want though.